src/neo4j_genai/retrievers/filters.py

```diff
@@ -0,0 +1,345 @@
+        if operator not in SUPPORTED_OPERATORS:
+            raise ValueError(
+                f"Invalid operator: {operator}. "
+                f"Expected one of {SUPPORTED_OPERATORS}"
+            )
+    else:  # if value is not dict, then we assume an equality operator
+        operator = OPERATOR_EQ
+        filter_value = value
+
+    # now everything is set, we can start and build the query
+    # special case for the BETWEEN operator that requires
+    # two tests (lower_bound <= value <= higher_bound)
+    if operator == OPERATOR_BETWEEN:
+        low, high = filter_value
+        param_name_low = param_store.add(field, low)
+        param_name_high = param_store.add(field, high)
+        query_snippet = (
+            f"${param_name_low} <= {DEFAULT_NODE_ALIAS}.`{field}` <= ${param_name_high}"
+        )
+        return query_snippet
+    # all the other operators are handled through their own classes:
+    native_op_class = COMPARISONS_TO_NATIVE[operator]
+    return _single_condition_cypher(
+        field, native_op_class, filter_value, param_store, node_alias
+    )
+
+
+def _construct_metadata_filter(
+    filter: dict[str, Any], param_store: ParameterStore, node_alias: str
+) -> str:
+    """Construct a metadata filter. This is a recursive function parsing the filter dict
+
+    Args:
+        filter: A dictionary representing the filter condition.
+        param_store: ParameterStore objet that will be updated in this function
+        node_alias: name of the node being filtered in the Cypher query
+
+    Returns:
+        str: the Cypher WHERE clause
+
+    NB: the param_store argument is mutable, it will be updated in this function
+    """
+
+    if not isinstance(filter, dict):
+        raise ValueError()
+    # if we have more than one entry, this is an implicit "AND" filter
+    if len(filter) > 1:
+        return _construct_metadata_filter(
+            {OPERATOR_AND: [{k: v} for k, v in filter.items()]}, param_store, node_alias
+        )
+    # The only operators allowed at the top level are $AND and $OR
+    # First check if an operator or a field
+    key, value = list(filter.items())[0]
+    if not key.startswith("$"):
+        # it's not an operator, must be a field
+        return _handle_field_filter(
+            key, filter[key], param_store, node_alias=node_alias
+        )
+
+    # Here we handle the $and and $or operators
+    if not isinstance(value, list):
+        raise ValueError(f"Expected a list, but got {type(value)} for value: {value}")
+    if key.lower() == OPERATOR_AND:
+        cypher_operator = " AND "
+    elif key.lower() == OPERATOR_OR:
+        cypher_operator = " OR "
+    else:
+        raise ValueError(f"Unsupported filter {filter}")
+    query = cypher_operator.join(
+        [
+            f"({ _construct_metadata_filter(el, param_store, node_alias)})"
+            for el in value
+        ]
+    )
+    return query
+
+
+def construct_metadata_filter(
```

Comment:
Minor but perhaps rename to `get_metadata_filter` or something else for this function? Given there is a private function with the same name